File: server/schema.py
from __future__ import annotations
from typing import Literal, Optional
from pydantic import BaseModel, Field


class NPCConversation(BaseModel):
    speechStyle: str = "warm, playful, and concise"
    favoriteTopic: str = "the cozy world around them"
    smallTalkFacts: list[str] = []
    openingChoices: list[str] = []
    clueLeadIn: str = "I did notice something that may help."
    animalSound: str = ""


class NPC(BaseModel):
    id: str
    name: str
    species: str
    appearance: str
    personality: str
    roomPosition: Literal["left", "right", "center"]
    ttsVoice: Optional[str] = None
    lines: NPCLines
    conversation: NPCConversation = NPCConversation()


class NPCLines(BaseModel):
    greeting: str
    hint: str
    idle: str


NPC.model_rebuild()


class Building(BaseModel):
    id: str
    name: str
    exteriorDescription: str
    sizeHint: Literal["small", "medium", "large"]
    isEnterable: bool
    interiorDescription: Optional[str] = None
    npcId: Optional[str] = None


class Hint(BaseModel):
    id: str
    npcId: str
    kind: Literal["attribute", "direction", "object"]
    text: str
    eliminatesLocationIds: list[str]


class ClueFacts(BaseModel):
    # Each value is a clause that can follow "Mama went somewhere ...".
    # Identical strings mean two candidates share that property.
    attribute: str
    direction: str
    object: str


class CandidateLocation(BaseModel):
    id: str
    name: str
    emoji: str
    description: str
    clueFacts: Optional[ClueFacts] = None


class Setting(BaseModel):
    name: str
    description: str
    decorations: list[str]


class Player(BaseModel):
    description: str
    accessory: str


class VisualDirection(BaseModel):
    layout: Literal["outdoor_hub", "interior_hub", "mixed_hub"] = "outdoor_hub"
    architecture: str = "storybook village"
    palette: list[str] = Field(default=["warm earth tones", "soft greens"], min_length=2, max_length=6)
    lighting: str = "soft daylight"
    atmosphere: str = "welcoming and playful"
    walkableSurface: str = "open paths and grass"
    imageStyle: str = "hand-painted children's storybook illustration"
    avoid: list[str] = []


class GameBible(BaseModel):
    title: str
    setting: Setting
    player: Player
    narratorIntro: str
    buildings: list[Building] = Field(min_length=5, max_length=7)
    npcs: list[NPC] = Field(min_length=3, max_length=3)
    hints: list[Hint] = Field(min_length=3, max_length=3)
    candidateLocations: list[CandidateLocation] = Field(min_length=4, max_length=4)
    finalLocationId: str
    finalLocationBuildingId: str
    reunionLine: str
    verification: str
    visualDirection: Optional[VisualDirection] = None


class DoorCell(BaseModel):
    col: int
    row: int


class Door(BaseModel):
    buildingId: str
    x: float
    y: float
    cell: DoorCell


class Spawn(BaseModel):
    col: int
    row: int


class ScreenAnnotation(BaseModel):
    grid: list[str]  # 16 strings of 24 chars each
    doors: list[Door]
    spawn: Spawn


class NpcBox(BaseModel):
    x: float
    y: float
    w: float
    h: float


class ExitPoint(BaseModel):
    x: float
    y: float


class RoomAnnotation(BaseModel):
    npcBox: NpcBox
    exit: ExitPoint


class SuggestedReply(BaseModel):
    text: str
    intent: Literal["ask_mama", "small_talk", "goodbye"]


class NPCModelReply(BaseModel):
    reply: str
    suggestedReplies: list[SuggestedReply]
    mood: Literal["cheerful", "curious", "helpful", "thoughtful", "excited"]
    conversationComplete: bool


class PreloadedNPCReply(NPCModelReply):
    audioUrl: Optional[str] = None


class RoomAssets(BaseModel):
    imageUrl: Optional[str] = None
    annotation: Optional[RoomAnnotation] = None


class GameAssets(BaseModel):
    streetUrl: Optional[str] = None
    outlineUrl: Optional[str] = None
    characterFrontUrl: Optional[str] = None
    characterSideUrl: Optional[str] = None
    rooms: dict[str, RoomAssets] = {}
    audio: dict[str, str] = {}
    npcOpeners: dict[str, PreloadedNPCReply] = {}
    npcMamaReplies: dict[str, PreloadedNPCReply] = {}
    npcMamaReplyVariants: dict[str, PreloadedNPCReply] = {}
    reunionPhotoUrl: Optional[str] = None


StageStatus = Literal["pending", "running", "done", "failed"]


class Game(BaseModel):
    id: str
    idea: str
    createdAt: str
    status: Literal["generating", "playable", "failed"]
    stages: dict[str, StageStatus] = {}
    bible: Optional[GameBible] = None
    assets: GameAssets = GameAssets()
    annotation: Optional[ScreenAnnotation] = None
    fromWarmPool: bool = False
    cozyVisuals: bool = False
    librarySourceId: Optional[str] = None

File: src/shared/schema.ts
import { z } from 'zod'

export const NPC = z.object({
  id: z.string(),
  name: z.string(),
  species: z.string(),
  appearance: z.string(),
  personality: z.string(),
  roomPosition: z.enum(['left', 'right', 'center']),
  ttsVoice: z.string().optional(),
  lines: z.object({
    greeting: z.string(),
    hint: z.string(),
    idle: z.string(),
  }),
  conversation: z.object({
    speechStyle: z.string(),
    favoriteTopic: z.string(),
    smallTalkFacts: z.array(z.string()),
    openingChoices: z.array(z.string()),
    clueLeadIn: z.string(),
    animalSound: z.string().nullish(),
  }).optional(),
})

export const Building = z.object({
  id: z.string(),
  name: z.string(),
  exteriorDescription: z.string(),
  sizeHint: z.enum(['small', 'medium', 'large']),
  isEnterable: z.boolean(),
  interiorDescription: z.string().optional(),
  npcId: z.string().nullable(),
})

export const Hint = z.object({
  id: z.string(),
  npcId: z.string(),
  kind: z.enum(['attribute', 'direction', 'object']),
  text: z.string(),
  eliminatesLocationIds: z.array(z.string()),
})

export const CandidateLocation = z.object({
  id: z.string(),
  name: z.string(),
  emoji: z.string(),
  description: z.string(),
  clueFacts: z.object({
    attribute: z.string(),
    direction: z.string(),
    object: z.string(),
  }).nullish(),
})

export const GameBible = z.object({
  title: z.string(),
  setting: z.object({
    name: z.string(),
    description: z.string(),
    decorations: z.array(z.string()),
  }),
  player: z.object({ description: z.string(), accessory: z.string() }),
  narratorIntro: z.string(),
  buildings: z.array(Building).min(5).max(7),
  npcs: z.array(NPC).length(3),
  hints: z.array(Hint).length(3),
  candidateLocations: z.array(CandidateLocation).length(4),
  finalLocationId: z.string(),
  finalLocationBuildingId: z.string(),
  reunionLine: z.string(),
  verification: z.string(),
  visualDirection: z.object({
    layout: z.enum(['outdoor_hub', 'interior_hub', 'mixed_hub']),
    architecture: z.string(),
    palette: z.array(z.string()).min(2).max(6),
    lighting: z.string(),
    atmosphere: z.string(),
    walkableSurface: z.string(),
    imageStyle: z.string(),
    avoid: z.array(z.string()),
  }).nullish(),
})

export const ScreenAnnotation = z.object({
  grid: z.array(z.string().length(24)).length(16),
  doors: z.array(z.object({
    buildingId: z.string(),
    x: z.number(),
    y: z.number(),
    cell: z.object({ col: z.number(), row: z.number() }),
  })),
  spawn: z.object({ col: z.number(), row: z.number() }),
})

export const RoomAnnotation = z.object({
  npcBox: z.object({ x: z.number(), y: z.number(), w: z.number(), h: z.number() }),
  exit: z.object({ x: z.number(), y: z.number() }),
})

export type GameBible = z.infer<typeof GameBible>
export type NPC = z.infer<typeof NPC>
export type Building = z.infer<typeof Building>
export type Hint = z.infer<typeof Hint>
export type CandidateLocation = z.infer<typeof CandidateLocation>
export type ScreenAnnotation = z.infer<typeof ScreenAnnotation>
export type RoomAnnotation = z.infer<typeof RoomAnnotation>

export interface SuggestedNPCReply {
  text: string
  intent: 'ask_mama' | 'small_talk' | 'goodbye'
}

export interface NPCChatResponse {
  sessionId: string
  reply: string
  suggestedReplies: SuggestedNPCReply[]
  mood: 'cheerful' | 'curious' | 'helpful' | 'thoughtful' | 'excited'
  conversationComplete: boolean
  audioUrl?: string | null
  clueGranted: boolean
  hintId?: string | null
}

export interface PreloadedNPCReply {
  reply: string
  suggestedReplies: SuggestedNPCReply[]
  mood: NPCChatResponse['mood']
  conversationComplete: boolean
  audioUrl?: string | null
}

export type Stage =
  | 'bible' | 'street' | 'outline' | 'hotspots' | 'character'
  | 'npc_openers' | `room_${string}` | 'tts' | 'reunion'

export type StageStatus = 'pending' | 'running' | 'done' | 'failed'

export interface Game {
  id: string
  idea: string
  createdAt: string
  status: 'generating' | 'playable' | 'failed'
  stages: Record<string, StageStatus>
  bible?: GameBible
  assets: {
    streetUrl?: string
    outlineUrl?: string
    characterFrontUrl?: string
    characterSideUrl?: string
    rooms: Record<string, { imageUrl?: string; annotation?: RoomAnnotation }>
    audio: Record<string, string>
    npcOpeners?: Record<string, PreloadedNPCReply>
    npcMamaReplies?: Record<string, PreloadedNPCReply>
    npcMamaReplyVariants?: Record<string, PreloadedNPCReply>
    reunionPhotoUrl?: string
  }
  annotation?: ScreenAnnotation
  fromWarmPool: boolean
  cozyVisuals?: boolean
  librarySourceId?: string
}

export interface AdventureSummary {
  id: string
  title: string
  settingName: string
  description: string
  idea: string
  previewUrl: string
  reunionPhotoUrl?: string | null
  createdAt: string
  emoji: string
}

export interface AdventurePage {
  items: AdventureSummary[]
  offset: number
  limit: number
  total: number
  hasMore: boolean
}

export interface SSEEvent {
  type: 'stage' | 'playable' | 'done' | 'failed'
  stage?: string
  status?: StageStatus
  meta?: Record<string, unknown>
  message?: string
}
